### ISSUE TITLE
trees.c unsigned integer overflow when using gzflush

### ISSUE THREAD

**Issue by nmoinvaz** [zlib-ng/zlib-ng#449]
I'm getting these errors when trying to use `gzflush` with Z_FINISH before `gzclose`:
```
trees.c:359:19: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'unsigned long'
trees.c:361:27: runtime error: unsigned integer overflow: 0 - 8 cannot be represented in type 'unsigned long'
trees.c:230:20: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'unsigned long'
trees.c:232:27: runtime error: unsigned integer overflow: 4294967288 + 8 cannot be represented in type 'unsigned long'
```
Here is a minimal example that should produce it:
```
#include "zbuild.h"
#ifdef ZLIB_COMPAT
#  include "zlib.h"
#else
#  include "zlib-ng.h"
#endif

#include <stdio.h>

#include <string.h>
#include <stdlib.h>
#include <inttypes.h>

int main(int argc, char *argv[])
{
    gzFile file = NULL;
    int written = 0;
    int err = 0;

    file = PREFIX(gzopen)("hello.gz", "wb");
    if (file == NULL) {
        fprintf(stderr, "gzopen error\n");
        return Z_STREAM_ERROR;
    }

    written = PREFIX(gzwrite)(file, "hello", 5);
    if (written < 0) {
        fprintf(stderr, "gzwrite err: %s\n", PREFIX(gzerror)(file, &err));
        return Z_STREAM_ERROR;
    }

    PREFIX(gzflush)(file, Z_FINISH);
    PREFIX(gzclose)(file);

    return Z_OK;
}
```

1. `gzflush` finishes the stream with `deflate/Z_FINISH` and calls `deflateReset` afterward. 
2. Then `gzclose` tries to also finish the stream with `deflate/Z_FINISH`. 
3. Because the previous call to `deflateReset` the stream doesn't see that it is already finished and tries to write a new block during `gzclose/deflate/Z_FINISH`. 
4. When it tries to write the trees for this new block, it throws these warnings most likely because no tree has been created because no new data has been read `strm->avail_in = 0` when it tries to write the new block.

Here is the fix that I came up with:

```
 gzwrite.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/gzwrite.c b/gzwrite.c
index 5b12805..58d2b73 100644
--- a/gzwrite.c
+++ b/gzwrite.c
@@ -54,6 +54,7 @@ static int gz_init(gz_state *state) {
 
     /* mark state as initialized */
     state->size = state->want;
+    state->reset = 0;
 
     /* initialize write buffer if compressing */
     if (!state->direct) {
@@ -122,7 +123,7 @@ static int gz_comp(gz_state *state, int flush) {
 
     /* if that completed a deflate stream, allow another to start */
     if (flush == Z_FINISH)
-        PREFIX(deflateReset)(strm);
+        state->reset = 1;
 
     /* all done, no errors */
     return 0;
@@ -177,6 +178,12 @@ static size_t gz_write(gz_state *state, void const *buf, size_t len) {
             return 0;
     }
 
+    /* reset deflate stream if previously flushed */
+    if (state->reset) {
+        PREFIX(deflateReset)(&state->strm);
+        state->reset = 0;
+    }
+
     /* for small len, copy to input buffer, otherwise compress directly */
     if (len < state->size) {
         /* copy to input buffer, compress when full */

```

Basically `gz_comp` sets a reset flag that upon the next `gzwrite` it will reset the deflate stream. Because in the code example above, `gzwrite` is never called between `gzflush` and `gzclose` it works and `gzclose` sees the deflate stream in the `FINISH_STATE`.

I'm thinking of submitting to madler/zlib.
